File: api/routers/mood.py
import logging
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from api.config import get_config
from api.services.mood_service import MoodService
from api.dependencies import get_current_user_id
from api.schemas.mood import (
    MoodCreate, MoodUpdate, MoodCreateResponse, MoodUpdateResponse,
    MoodLogCreate, MoodLogAddResponse, MoodLogDeleteResponse,
)

logger = logging.getLogger(__name__)


def create_mood_router(mood_service: MoodService, fitness_service=None) -> APIRouter:
    router = APIRouter()

    @router.post("/mood", status_code=201, response_model=MoodCreateResponse)
    def create_mood_entry(
        body: MoodCreate,
        user_id: int = Depends(get_current_user_id),
    ):
        try:
            result = mood_service.create_mood_entry(
                user_id,
                body.date,
                body.mood,
                body.content,
                body.time,
                body.selected_options,
            )
            return {
                "status": "success",
                "entry_id": result["entry_id"],
                "new_achievements": result["new_achievements"],
                "message": "Mood entry created successfully",
            }
        except ValueError as e:
            raise HTTPException(status_code=400, detail=str(e))
        except Exception as e:
            logger.error(f"Create mood entry error: {e}")
            raise HTTPException(status_code=500, detail=str(e))

    def _enrich_with_fitness(entries, user_id: int):
        """Attach fitness data to entries when Google Health is enabled."""
        if not fitness_service:
            return entries
        try:
            cfg = get_config()
            if not cfg.ENABLE_GOOGLE_HEALTH:
                return entries
            dates = [e["date"] for e in entries] if isinstance(entries, list) else [entries["date"]]
            fitness_by_date = fitness_service.get_data_for_dates(user_id, dates)
            if isinstance(entries, list):
                for entry in entries:
                    entry["fitness"] = fitness_by_date.get(entry["date"], [])
            else:
                entries["fitness"] = fitness_by_date.get(entries["date"], [])
        except Exception as exc:
            logger.warning("Failed to attach fitness data: %s", exc)
        return entries

    @router.get("/moods")
    def get_mood_entries(
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        user_id: int = Depends(get_current_user_id),
    ):
        try:
            if start_date and end_date:
                entries = mood_service.get_entries_by_date_range(user_id, start_date, end_date)
            else:
                entries = mood_service.get_all_entries(user_id)
            return _enrich_with_fitness(entries, user_id)
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.get("/mood/{entry_id}")
    def get_mood_entry(entry_id: int, user_id: int = Depends(get_current_user_id)):
        try:
            entry = mood_service.get_entry_by_id(user_id, entry_id)
            if entry is None:
                raise HTTPException(status_code=404, detail="Entry not found")
            return _enrich_with_fitness(entry, user_id)
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.put("/mood/{entry_id}", response_model=MoodUpdateResponse)
    def update_mood_entry(
        entry_id: int,
        body: MoodUpdate,
        user_id: int = Depends(get_current_user_id),
    ):
        try:
            fields = body.model_dump(exclude_unset=True)
            if not fields:
                raise HTTPException(status_code=400, detail="No update fields provided")

            updated = mood_service.update_entry(
                user_id,
                entry_id,
                mood=fields.get('mood'),
                content=fields.get('content'),
                date=fields.get('date'),
                time=fields.get('time'),
                selected_options=fields.get('selected_options'),
            )
            if updated is None:
                raise HTTPException(status_code=404, detail="Entry not found or no changes made")
            return {
                "status": "success",
                "message": "Mood entry updated successfully",
                "entry": updated,
            }
        except HTTPException:
            raise
        except ValueError as e:
            raise HTTPException(status_code=400, detail=str(e))
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.delete("/mood/{entry_id}")
    def delete_mood_entry(entry_id: int, user_id: int = Depends(get_current_user_id)):
        try:
            success = mood_service.delete_entry(user_id, entry_id)
            if not success:
                raise HTTPException(status_code=404, detail="Entry not found")
            return {"status": "success", "message": "Mood entry deleted successfully"}
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.get("/statistics")
    def get_mood_statistics(user_id: int = Depends(get_current_user_id)):
        try:
            return mood_service.get_statistics(user_id)
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.get("/streak")
    def get_current_streak(user_id: int = Depends(get_current_user_id)):
        try:
            streak = mood_service.get_current_streak(user_id)
            return {
                "current_streak": streak,
                "message": f'Current streak: {streak} day{"s" if streak != 1 else ""}',
            }
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.get("/mood/{entry_id}/selections")
    def get_entry_selections(entry_id: int, user_id: int = Depends(get_current_user_id)):
        try:
            return mood_service.get_entry_selections(user_id, entry_id)
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.post("/mood/{entry_id}/logs", status_code=201, response_model=MoodLogAddResponse)
    def add_mood_log(
        entry_id: int,
        body: MoodLogCreate,
        user_id: int = Depends(get_current_user_id),
    ):
        try:
            result = mood_service.add_mood_log(user_id, entry_id, body.mood)
            return {
                "status": "success",
                "log_id": result["log_id"],
                "mood": result["mood"],
                "logged_at": result["logged_at"],
                "updated_entry_mood": result["updated_entry_mood"],
                "message": "Mood log added",
            }
        except ValueError as e:
            raise HTTPException(status_code=404, detail=str(e))
        except Exception as e:
            logger.error(f"Add mood log error: {e}")
            raise HTTPException(status_code=500, detail=str(e))

    @router.get("/mood/{entry_id}/logs")
    def get_mood_logs(entry_id: int, user_id: int = Depends(get_current_user_id)):
        try:
            return mood_service.get_mood_logs(user_id, entry_id)
        except ValueError as e:
            raise HTTPException(status_code=404, detail=str(e))
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @router.delete("/mood/{entry_id}/logs/{log_id}", response_model=MoodLogDeleteResponse)
    def delete_mood_log(
        entry_id: int,
        log_id: int,
        user_id: int = Depends(get_current_user_id),
    ):
        try:
            result = mood_service.delete_mood_log(user_id, entry_id, log_id)
            if result is None:
                raise HTTPException(status_code=404, detail="Log not found")
            return {
                "status": "success",
                "updated_entry_mood": result["updated_entry_mood"],
                "message": "Mood log deleted",
            }
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    return router

File: functions/_shared/routes/mood.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { z } from 'zod';
import { HTTPException } from 'hono/http-exception';
import type { AppEnv } from '../lib/env';
import { createDb } from '../db/client';
import { moodCreateSchema, moodLogCreateSchema, moodUpdateSchema } from '@shared/schemas/mood';
import {
  ValidationError,
  addMoodLog,
  createMoodEntry,
  deleteEntry,
  deleteMoodLog,
  getCurrentStreak,
  getEntryById,
  getEntrySelections,
  getEntrySliderValues,
  getHydratedEntries,
  getMoodLogs,
  getStatistics,
  updateEntry,
} from '../services/mood-service';

const moodQuerySchema = z.object({
  start_date: z.string().optional(),
  end_date: z.string().optional(),
});

const entryIdParamSchema = z.object({ entry_id: z.coerce.number().int() });
const logIdParamSchema = z.object({ entry_id: z.coerce.number().int(), log_id: z.coerce.number().int() });

export const moodRoutes = new Hono<AppEnv>();

moodRoutes.post('/mood', zValidator('json', moodCreateSchema), async (c) => {
  const userId = c.get('userId');
  const body = c.req.valid('json');
  const db = createDb(c.env.DB);

  try {
    const result = await createMoodEntry(db, userId, body);
    return c.json(
      {
        status: 'success',
        entry_id: result.entry_id,
        new_achievements: result.new_achievements,
        message: 'Mood entry created successfully',
      },
      201
    );
  } catch (err) {
    if (err instanceof ValidationError) {
      throw new HTTPException(400, { message: err.message });
    }
    throw err;
  }
});

moodRoutes.get('/moods', zValidator('query', moodQuerySchema), async (c) => {
  const userId = c.get('userId');
  const { start_date: startDate, end_date: endDate } = c.req.valid('query');
  const db = createDb(c.env.DB);

  const hydratedEntries = await getHydratedEntries(db, userId, {
    startDate: startDate ?? undefined,
    endDate: endDate ?? undefined,
  });

  return c.json(hydratedEntries);
});

moodRoutes.get('/mood/:entry_id', zValidator('param', entryIdParamSchema), async (c) => {
  const userId = c.get('userId');
  const { entry_id: entryId } = c.req.valid('param');
  const db = createDb(c.env.DB);

  const entry = await getEntryById(db, userId, entryId);
  if (!entry) {
    throw new HTTPException(404, { message: 'Entry not found' });
  }
  return c.json(entry);
});

moodRoutes.put(
  '/mood/:entry_id',
  zValidator('param', entryIdParamSchema),
  zValidator('json', moodUpdateSchema),
  async (c) => {
    const userId = c.get('userId');
    const { entry_id: entryId } = c.req.valid('param');
    const body = c.req.valid('json');
    const db = createDb(c.env.DB);

    if (Object.keys(body).length === 0) {
      throw new HTTPException(400, { message: 'No update fields provided' });
    }

    try {
      const updated = await updateEntry(db, userId, entryId, body);
      if (!updated) {
        throw new HTTPException(404, { message: 'Entry not found or no changes made' });
      }
      return c.json({ status: 'success', message: 'Mood entry updated successfully', entry: updated });
    } catch (err) {
      if (err instanceof HTTPException) throw err;
      if (err instanceof ValidationError) {
        throw new HTTPException(400, { message: err.message });
      }
      throw err;
    }
  }
);

moodRoutes.delete('/mood/:entry_id', zValidator('param', entryIdParamSchema), async (c) => {
  const userId = c.get('userId');
  const { entry_id: entryId } = c.req.valid('param');
  const db = createDb(c.env.DB);

  const success = await deleteEntry(db, userId, entryId);
  if (!success) {
    throw new HTTPException(404, { message: 'Entry not found' });
  }
  return c.json({ status: 'success', message: 'Mood entry deleted successfully' });
});

moodRoutes.get('/statistics', async (c) => {
  const userId = c.get('userId');
  const db = createDb(c.env.DB);
  return c.json(await getStatistics(db, userId));
});

moodRoutes.get('/streak', async (c) => {
  const userId = c.get('userId');
  const db = createDb(c.env.DB);
  const streak = await getCurrentStreak(db, userId);
  return c.json({
    current_streak: streak,
    message: `Current streak: ${streak} day${streak !== 1 ? 's' : ''}`,
  });
});

moodRoutes.get('/mood/:entry_id/selections', zValidator('param', entryIdParamSchema), async (c) => {
  const userId = c.get('userId');
  const { entry_id: entryId } = c.req.valid('param');
  const db = createDb(c.env.DB);

  const entry = await getEntryById(db, userId, entryId);
  if (!entry) return c.json([]);
  return c.json(await getEntrySelections(db, entryId));
});

moodRoutes.get('/mood/:entry_id/slider-values', zValidator('param', entryIdParamSchema), async (c) => {
  const userId = c.get('userId');
  const { entry_id: entryId } = c.req.valid('param');
  const db = createDb(c.env.DB);

  const entry = await getEntryById(db, userId, entryId);
  if (!entry) return c.json([]);
  return c.json(await getEntrySliderValues(db, entryId));
});

moodRoutes.post(
  '/mood/:entry_id/logs',
  zValidator('param', entryIdParamSchema),
  zValidator('json', moodLogCreateSchema),
  async (c) => {
    const userId = c.get('userId');
    const { entry_id: entryId } = c.req.valid('param');
    const { mood } = c.req.valid('json');
    const db = createDb(c.env.DB);

    try {
      const result = await addMoodLog(db, userId, entryId, mood);
      return c.json(
        {
          status: 'success',
          log_id: result.log_id,
          mood: result.mood,
          logged_at: result.logged_at,
          updated_entry_mood: result.updated_entry_mood,
          message: 'Mood log added',
        },
        201
      );
    } catch (err) {
      if (err instanceof ValidationError) {
        throw new HTTPException(404, { message: err.message });
      }
      throw err;
    }
  }
);

moodRoutes.get('/mood/:entry_id/logs', zValidator('param', entryIdParamSchema), async (c) => {
  const userId = c.get('userId');
  const { entry_id: entryId } = c.req.valid('param');
  const db = createDb(c.env.DB);

  try {
    return c.json(await getMoodLogs(db, userId, entryId));
  } catch (err) {
    if (err instanceof ValidationError) {
      throw new HTTPException(404, { message: err.message });
    }
    throw err;
  }
});

moodRoutes.delete('/mood/:entry_id/logs/:log_id', zValidator('param', logIdParamSchema), async (c) => {
  const userId = c.get('userId');
  const { entry_id: entryId, log_id: logId } = c.req.valid('param');
  const db = createDb(c.env.DB);

  const result = await deleteMoodLog(db, userId, entryId, logId);
  if (result === null) {
    throw new HTTPException(404, { message: 'Log not found' });
  }
  return c.json({ status: 'success', updated_entry_mood: result, message: 'Mood log deleted' });
});
